pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,208 @@
+package sa
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/openshift/origin/pkg/cmd/util"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/serviceaccounts"
+	osautil "github.com/openshift/origin/pkg/serviceaccounts/util"
+)
+
+const (
+	NewServiceAccountTokenRecommendedName = "new-token"
+
+	newServiceAccountTokenShort = `Generate a new service account API token for a service account.`
+
+	newServiceAccountTokenLong = `
+Generate a new service account API token for a service account.
+
+Service account API tokens are used by service accounts for privileged access to the API. This command will generate a new
+token, which could be used to compartmentalize service account actions by executing them with distinct tokens, to rotate out
+pre-existing token on the service account, or for use by an external client.
+`
+
+	newServiceAccountTokenUsage = `%s SA-NAME`
+
+	newServiceAccountTokenExamples = `  # Generate a new token and for service account 'custom-sa'
+  $ %[1]s 'custom-sa'
+`
+)
+
+type NewServiceAccountTokenOptions struct {
+	SAName        string
+	Namespace     string
+	SAClient      unversioned.ServiceAccountsInterface
+	SecretsClient unversioned.SecretsInterface
+
+	TimeoutSeconds int64
+
+	Out io.Writer
+	Err io.Writer
+}
+
+func NewCommandNewServiceAccountToken(name, fullname string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &NewServiceAccountTokenOptions{
+		Out: out,
+		Err: os.Stderr,
+	}
+
+	newServiceAccountTokenCommand := &cobra.Command{
+		Use:     fmt.Sprintf(newServiceAccountTokenUsage, name),
+		Short:   newServiceAccountTokenShort,
+		Long:    newServiceAccountTokenLong,
+		Example: fmt.Sprintf(newServiceAccountTokenExamples, fullname),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := options.Complete(args, f, cmd); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+
+			cmdutil.CheckErr(options.Run())
+		},
+	}
+
+	newServiceAccountTokenCommand.Flags().Int64Var(&options.TimeoutSeconds, "timeout", 100, "the maximum time allowed to generate a token, in seconds")
```

Comment:
> We don't use time.Duration flags anywhere else but sure we can start.

We use it in our config because it keeps people from needing to know the scale used for a field.